design/control-data-plane-separation/design.md

```diff
@@ -0,0 +1,770 @@
+  intended recipient of the token.
+- Object-bound: Bound tokens are bound to the Pod.
+
+The TokenReview API only considers a bound token to be valid if the token is not expired, the audience of the token
+matches the audience specified in the TokenReview API request, and the Pod that the token is bound to is still present
+and running.
+
+Bound tokens expire, and are written to the filesystem by the kubelet. While bound tokens are more secure than
+long-lived tokens, the agent needs to be modified to use them. The agent would need to be able to reload the tokens from
+the filesystem periodically. That would require the following changes in the agent code:
+
+- Add a new configuration option to specify the path to the token file. Currently, the agent supports reading the token
+  from an environment variable or the configuration file, not from a file.
+- Modify the gRPC client to fetch the token from a file before connecting to the control plane. Currently, the token is
+  loaded on start-up and never refreshed. If the agent reconnects to the control plane, it will use the same token
+  provided on start-up.
+
+I propose that we initially use a long-lived token for the agent. This will allow us to get the agent up and running
+quickly. We can then add support for bound tokens in a future release.
+
+To create the long-lived token, we will provide the following manifest:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: nginx-agent-token-secret
+  annotations:
+    kubernetes.io/service-account.name: nginx-agent
+type: kubernetes.io/service-account-token
+```
+
+And expose the token as an environment variable in the agent container:
+
+```yaml
+    env:
+    - name: NMS_SERVER_TOKEN
+      valueFrom:
+        secretKeyRef:
+          name: nginx-agent-token-secret
+          key: token
+```
+
+The agent will load the token from the `$NSM_SERVER_TOKEN` environment variable and add it to the `Authorization`
+header of the gRPC request when connecting to the control plane.
+
+For a good comparison of long-lived and bound tokens, see [this blog post][bound-token-gke].
+
+[token-review]: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/
+
+[bound-token-gke]: https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-bound-service-account-tokens
+
+[token-request]: https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-request-v1/
+
+[projected-volume]: https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#bound-service-account-token-volume
+
+## Agent Registration
+
+![Agent Connect Response](./connect-response.png)
+
+The agent will send an [`AgentConnectRequest`][connect] command to the control plane over the`CommandChannel` to
+register with the control plane. This command includes the agent’s system ID and nginx ID(s) which the control plane
```

Comment:
If the agent Pod restarts it will re-register itself with the control plane. We will probably need to have some connection manager that associates IDs with connections and can remove the agent entry if the connection is dropped. 

The system ID is generated using the container ID and the nginx ID(s) is computed as a checksum using the nginx process path, conf path, and prefix.

Reply:
If the control plane restarts the agents connections to the control plane will be dropped and they will attempt to reconnect in a loop.